src/ray/core_worker/core_worker.cc

```diff
@@ -2458,29 +2456,60 @@ void CoreWorker::HandleSubscribeForObjectEviction(
   // Returns true if the object was present and the callback was added. It might have
   // already been evicted by the time we get this request, in which case we should
   // respond immediately so the raylet unpins the object.
-  if (!reference_counter_->SetDeleteCallback(object_id, respond)) {
+  if (!reference_counter_->SetDeleteCallback(object_id, unpin_object)) {
     // If the object is already evicted (callback cannot be set), unregister the
-    // subscription.
-    object_status_publisher_->UnregisterSubscription(
-        rpc::ChannelType::WORKER_OBJECT_EVICTION, subscriber_node_id, object_id.Binary());
+    // subscription & publish the message so that the subscriber knows it.
+    unpin_object(object_id);
     std::ostringstream stream;
     stream << "Reference for object " << object_id << " has already been freed.";
     RAY_LOG(DEBUG) << stream.str();
-    send_reply_callback(Status::NotFound(stream.str()), nullptr, nullptr);
-  } else {
-    send_reply_callback(Status::OK(), nullptr, nullptr);
+  }
+}
+
+void CoreWorker::ProcessPubsubCommands(const Commands &commands,
+                                       const NodeID &subscriber_id) {
+  for (const auto &command : commands) {
+    if (command.has_unsubscribe_message()) {
+      object_status_publisher_->UnregisterSubscription(command.channel_type(),
+                                                       subscriber_id, command.key_id());
+      continue;
+    }
+
+    RAY_CHECK(command.has_subscribe_message())
```

Comment:
I'd prefer 
```
if () {
} else if{
} else {
LOG(FATAL) << "Unsupported type"
}
```

In this way, later if you add extra things, it can be added easily.